shared/gcs/gcs.go

```diff
@@ -19,96 +19,161 @@ limitations under the License.
 package gcs
 
 import (
-	"bufio"
 	"context"
-	"fmt"
 	"io/ioutil"
 	"log"
-	"strconv"
-	"strings"
+	"path"
+	"os"
+	"io"
 
 	"cloud.google.com/go/storage"
 	"google.golang.org/api/option"
-)
-
-const (
-	bucketName = "knative-prow"
-	latest     = "/latest-build.txt"
+	"google.golang.org/api/iterator"
 )
 
 var client *storage.Client
 
-func createStorageClient(ctx context.Context, sa string) error {
+// Authenticate explicitly sets up authentication for the rest of run
+func Authenticate(ctx context.Context, sa string) error {
 	var err error
 	client, err = storage.NewClient(ctx, option.WithCredentialsFile(sa))
 	return err
 }
 
-func createStorageObject(filename string) *storage.ObjectHandle {
-	return client.Bucket(bucketName).Object(filename)
+// Exist checks if path exist under gcs bucket
+func Exist(ctx context.Context, bucketName, filePath string) bool {
+	handle := createStorageObject(bucketName, filePath)
+	if _, err := handle.Attrs(ctx); nil != err {
+		return false
+	}
+	return true
+}
+
+// List child under prefix, use delim for skipping some files.
+// This function gets all child files recursively under given prefix,
+// then filter out filenames containing giving delim.
+// If delim is empty string, returns all files but not directories,
+// if delim is "/", returns all direct children, including both files and directories.
+// see https://godoc.org/cloud.google.com/go/storage#Query
+func List(ctx context.Context, bucketName, prefix, delim string) []string {
+	var filePaths []string
+	objsAttrs := getObjectsAttrs(ctx, bucketName, prefix, delim)
+	for _, attrs := range objsAttrs {
+		filePaths = append(filePaths, path.Join(attrs.Prefix, attrs.Name))
+	}
+	return filePaths
+}
+
+// ListDirectChildren lists direct children paths(including files and directories).
+// Note: to avoid including unnecessary directories, prefix must end with "/".
```

Comment:
since the desired behavior of this function is getting children inside given `directory` but not `directory_foo`, enforced adding trailing slash to storagePath inside this function